finality-aleph/src/new_network/connection_manager/discovery.rs

```diff
@@ -0,0 +1,690 @@
+use crate::{
+    new_network::{
+        connection_manager::{Authentication, Multiaddr, SessionHandler},
+        DataCommand, PeerId, Protocol,
+    },
+    NodeIndex, SessionId,
+};
+use codec::{Decode, Encode};
+use std::{
+    collections::{HashMap, HashSet},
+    time,
+};
+
+/// Messages used for discovery and authentication.
+#[derive(Clone, Debug, Encode, Decode)]
+pub enum DiscoveryMessage {
+    // Contains the id of the broadcast, to avoid spam.
+    AuthenticationBroadcast(Authentication),
+    // Requests always contain own authentication, to avoid asymetric trust.
+    Request(Vec<NodeIndex>, Authentication),
+    Authentications(Vec<Authentication>),
+}
+
+impl DiscoveryMessage {
+    pub fn session_id(&self) -> SessionId {
+        use DiscoveryMessage::*;
+        match self {
+            AuthenticationBroadcast((auth_data, _)) => auth_data.session(),
+            Request(_, (auth_data, _)) => auth_data.session(),
+            Authentications(auths) => match auths.get(0) {
+                Some((auth_data, _)) => auth_data.session(),
+                None => SessionId(0), // Broken message anyway, value doesn't matter.
+            },
+        }
+    }
+}
+
+/// Handles creating and responding to discovery messages.
+pub struct Discovery {
+    cooldown: time::Duration,
+    last_broadcast: HashMap<NodeIndex, time::Instant>,
+    last_response: HashMap<NodeIndex, time::Instant>,
+    requested_authentications: HashMap<NodeIndex, HashSet<NodeIndex>>,
+    next_query: NodeIndex,
+}
+
+fn authentication_broadcast(authentication: Authentication) -> (DiscoveryMessage, DataCommand) {
+    (
+        DiscoveryMessage::AuthenticationBroadcast(authentication),
+        DataCommand::Broadcast,
+    )
+}
+
+fn request(
+    missing_authorities: Vec<NodeIndex>,
+    authentication: Authentication,
+    peer_id: PeerId,
+) -> (DiscoveryMessage, DataCommand) {
+    (
+        DiscoveryMessage::Request(missing_authorities, authentication),
+        DataCommand::SendTo(peer_id, Protocol::Generic),
+    )
+}
+
+fn response(
+    authentications: Vec<Authentication>,
+    peer_id: PeerId,
+) -> (DiscoveryMessage, DataCommand) {
+    (
+        DiscoveryMessage::Authentications(authentications),
+        DataCommand::SendTo(peer_id, Protocol::Generic),
+    )
+}
+
+impl Discovery {
+    /// Create a new discovery handler with the given response/broadcast cooldown.
+    pub fn new(cooldown: time::Duration) -> Self {
+        Discovery {
+            cooldown,
+            last_broadcast: HashMap::new(),
+            last_response: HashMap::new(),
+            requested_authentications: HashMap::new(),
```

Comment:
can we say anything about (upper bound for) the size of these containers?

Reply:
Yup, at most one item per node in session (we only add stuff that is in some way authenticated). The instants are probably small, the hashsets of nodeindices are again bound by the number of nodes in the session.

Whoops, the latter actually weren't, added a check.

Reply:
but over time, the number of entries can be as large as the number of all nodes that could reach us? should we be careful with that?

Reply:
Huh? I'm kind of confused, the number of nodes is the number of committee members in this case, and that is bounded. Discovery exists per-session, so it will get deleted after a session passes.